32-bit build test fails when AVX2 is available

When running on a 64-bit host, a 32-bit build fails tests because AVX2 is available and something does not handle this right:
```
$ go version
go version go1.12.2 linux/amd64

$ GOARCH=amd64 go test
PASS
ok  	github.com/minio/sha256-simd	0.002s

$ GOARCH=386 go test
--- FAIL: TestGolden (0.00s)
    sha256_test.go:2227: AVX2: Sum256 function: sha256() = 6a09e667bb67ae853c6ef372a54ff53a510e527f9b05688c1f83d9ab5be0cd19 want e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
FAIL
exit status 1
FAIL	github.com/minio/sha256-simd	0.001s
```
Running on Fedora 30.

--- a/sha256_test.go
+++ b/sha256_test.go
@@ -52,6 +52,7 @@ package sha256
 import (
 	"encoding/hex"
 	"fmt"
+	"runtime"
 	"strings"
 	"testing"
 )
@@ -2210,6 +2211,25 @@ var golden = []sha256Test{
 func TestGolden(t *testing.T) {
 	blockfuncSaved := blockfunc
 
+	defer func() {
+		blockfunc = blockfuncSaved
+	}()
+
+	if true {
+		blockfunc = blockfuncGeneric
+		for _, g := range golden {
+			s := fmt.Sprintf("%x", Sum256([]byte(g.in)))
+			if Sum256([]byte(g.in)) != g.out {
+				t.Fatalf("Generic: Sum256 function: sha256(%s) = %s want %s", g.in, s, hex.EncodeToString(g.out[:]))
+			}
+		}
+	}
+
+	if runtime.GOARCH == "386" || runtime.GOARCH == "arm" {
+		// doesn't support anything but the generic version.
+		return
+	}
+
 	if sha && ssse3 && sse41 {
 		blockfunc = blockfuncSha
 		for _, g := range golden {
@@ -2246,17 +2266,6 @@ func TestGolden(t *testing.T) {
 			}
 		}
 	}
-	if true {
-		blockfunc = blockfuncGeneric
-		for _, g := range golden {
-			s := fmt.Sprintf("%x", Sum256([]byte(g.in)))
-			if Sum256([]byte(g.in)) != g.out {
-				t.Fatalf("Generic: Sum256 function: sha256(%s) = %s want %s", g.in, s, hex.EncodeToString(g.out[:]))
-			}
-		}
-	}
-
-	blockfunc = blockfuncSaved
 }
 
 func TestSize(t *testing.T) {

